### DISCUSSION TITLE
Improve VPN Model to Allow for Multi-Select Options

### DISCUSSION THREAD

**Discussion by markh0338** [netbox-community/netbox#14820] [General]
I wrote about this in issue #14790 but now creating a discussion since it sounds like this will not be simple to implement. It seems to be common (maybe not best) practice to configure multiple encryption/integrity algorithms in a single isakmp/ikev2 proposals. However, the new vpn model in netbox only allows for a single option to be selected per proposal.

Something like this cannot be documented currently in version v3.7.0 of netbox:
```
crypto ikev2 proposal ike-proposal-1
 encryption aes-cbc-256 aes-cbc-128
 integrity sha512 sha256
 group 21 20 19
```

Has anyone else ran into this limitation? Or found a different way to work around it?

**Comment by jcollie** [comment]
Yes, I ran into it with the very first IPSec VPN that I tried to document. I would say that this is an overwhelmingly common practice. In theory you could set up separate policies for all of the combinations but that quickly becomes infeasible. Just the above example would require 12 entries in NetBox to duplicate. Then you have the reverse problem of trying to generate a VPN config from what's stored in NetBox.

**Comment by jeremystretch** [comment]
> I would say that this is an overwhelmingly common practice.

No one raised it as an issue either during the development of #9816 or during the month-long beta evaluation period, so it must be not be too common.

Had it been raised as a limitation during the beta period, we could have revisited the data model. However, now that v3.7 has been released, any breaking changes will need to wait for a future minor version bump. This is why we beg people to participate in the evaluation period.

**Comment by jcollie** [comment]
Jeremy,

Expecting a large number of people to find some random issue in the bug tracker and follow the discussion in that issue is not reasonable, especially if they aren't looking for it.

Expecting a large number of people to participate in a beta test over the winter holidays is not reasonable.

Limitations of the model were brought up by @jronnblom before the beta was released but were ignored.
https://github.com/netbox-community/netbox/issues/9816#issuecomment-1825322367

I think that this issue and others highlights a deficiency in NetBox's current development model. Large feature changes/additions like this are developed very quickly with discussions that are buried in issues and then design decisions are locked down way too quickly. At the minimum feature additions/changes like this should have a discussion associated with them, probably with a blog post explaining the feature proposal and a link for further discussion.

It would help also if the beta announcements would contain reminders about how to access the beta demo site or how to install the beta locally for testing.

**Reply by jeremystretch** [reply]
> Expecting a large number of people to participate in a beta test over the winter holidays is not reasonable.

Again, this was a 30-day beta period, the same as for every minor release we have, and the FR had been tagged for the v3.7 release for months prior. If you think that's unreasonable, we'll simply have to disagree.

The simple fact is that if you opt not to participate in the development of the project, the impact of your feedback after the fact will be limited.

**Comment by antoinekh** [comment]
Sorry I'm late to the game but I just found the post after trying all solution to avoid having to put dozen of Ike proposal just to simulate the proposal that I have in reality.
On my Paloalto I need a large IKE profile with 4 encryption, 3 integrity, 5 group. And with the current implementation I have to create 60 proposals for that and extract data from each of them every time I want to push the profile to a device.

So if anyone has a better solution, I will use it. Has anyone tried to modify VPN model to allow multiple selections?

**Comment by antoinekh** [comment]
Another solution I just tried and that seems to work, is not to use IKE proposals.
I'm going to create a dummy one because it's necessary.

I'm going to use custom fileds on the IKE policy, which is the closest thing to production.

it gives me something like this:
![image](https://github.com/netbox-community/netbox/assets/52744345/1d7670fd-2bb7-4086-8042-b1f93af38833)
![image](https://github.com/netbox-community/netbox/assets/52744345/2c37f7d0-4ae3-4603-8af2-af5b9c469925)

And the data to build the VPN are directly and completely accessible via the API
![image](https://github.com/netbox-community/netbox/assets/52744345/f191cf07-8082-4b67-9719-c6fa39b49a7a)


It's not the cleanest solution, but at least it represents production
You just need to manage custom field choices properly.

**Reply by jo-bzh** [reply]
It seems the only solution to avoid a huge amount of IKE Proposals, most streamlined approach.

**Comment by skyefugate** [comment]
Heyo! Having a similar issue with getting multiple tied to a proposal. Anyone found a workaround for 4.2.3?

**Reply by antoinekh** [reply]
Yes with custom fields. Far from perfect but it works. You can check my previous message
